src/Microsoft.SqlTools.ServiceLayer/SqlAssessment/SqlAssessmentService.cs

```diff
@@ -0,0 +1,512 @@
+
+                ConnectParams connectParams = new ConnectParams
+                {
+                    OwnerUri = randomUri,
+                    Connection = connInfo.ConnectionDetails,
+                    Type = ConnectionType.Default
+                };
+
+                if(!connInfo.TryGetConnection(ConnectionType.Default, out var connection))
+                {
+                    await requestContext.SendError(SR.SqlAssessmentConnectingError);
+                }
+
+                var workTask = CallAssessmentEngine<TResult>(
+                        requestParams,
+                        connectParams,
+                        randomUri,
+                        assessmentFunction)
+                    .ContinueWith(async tsk =>
+                    {
+                        await requestContext.SendResult(tsk.Result);
+                    });
+
+                ActiveRequests.TryAdd(randomUri, workTask);
+            }
+            catch (Exception ex)
+            {
+                if (ex is StackOverflowException || ex is OutOfMemoryException)
+                {
+                    throw;
+                }
+
+                await requestContext.SendError(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// This function obtains a live connection, then calls
+        /// an assessment operation specified by <paramref name="assessmentFunc"/>
+        /// </summary>
+        /// <typeparam name="TResult">
+        /// SQL Assessment result item type.
+        /// </typeparam>
+        /// <param name="requestParams">
+        /// Request parameters passed from the host.
+        /// </param>
+        /// <param name="connectParams">
+        /// Connection parameters used to identify and access the target.
+        /// </param>
+        /// <param name="taskUri">
+        /// An URI identifying the request task to enable concurrent execution.
+        /// </param>
+        /// <param name="assessmentFunc">
+        /// A function performing assessment operation for given target.
+        /// </param>
+        /// <returns>
+        /// Returns <see cref="AssessmentResult{TResult}"/> for given target.
+        /// </returns>
+        internal async Task<AssessmentResult<TResult>> CallAssessmentEngine<TResult>(
+            AssessmentParams requestParams,
+            ConnectParams connectParams,
+            string taskUri,
+            Func<SqlObjectLocator, Task<List<TResult>>> assessmentFunc)
+            where TResult : AssessmentItemInfo
+
+        {
+            var result = new AssessmentResult<TResult>
+            {
+                ApiVersion = ApiVersion
+            };
+
+            await ConnectionService.Connect(connectParams);
+
+            var connection = await ConnectionService.Instance.GetOrOpenConnection(taskUri, ConnectionType.Query);
+
+            try
+            {
+                const string PlatformQueryText = @"SELECT [host_platform] FROM sys.dm_os_host_info";
+

```

Comment:
Why not add this to the other helper method (with the same logic to check for server version first)? Again this seems like a generally useful thing to have

Reply:
Moved to `ReliableConnectionHelper.GetServerHostInfo`.